contracts/pools/weighted-oracle/WeightedOracleMath.sol

```diff
@@ -14,60 +14,88 @@
-        uint256 balanceB
-    ) internal pure returns (int256 invariantLn) {
-        //We can cast weights and balances to int256 becuase they are always lower than the max int256.
-        int256 term1 = int256(normalizedWeightA).mul(SignedFixedPoint.ln(int256(balanceA)));
-        int256 term2 = int256(normalizedWeightB).mul(SignedFixedPoint.ln(int256(balanceB)));
+    /**
+     * @dev Returns the natural logarithm of `value`, dropping most of the decimal places to arrive at a value that,
+     * when passed to `_fromLowResLog`, will have a maximum relative error of ~0.05% compared to `value`.
+     *
+     * Values returned from this function should not be mixed with other fixed-point values (as they have different
+     * number of digits), but can be added or subtracted. Use `_fromLowResLog` to undo this process and return to an
+     * 18 decimal places fixed point value.
+     *
+     * Because so much precision is lost, the logarithmic values can be stored using much fewer bits than the original
+     * value required.
+     */
+    function _toLowResLog(uint256 value) internal pure returns (int256) {
+        int256 ln = LogExpMath.ln(int256(value));
 
-        invariantLn = term1.add(term2) / _INVARIANT_COMPRESSION_FACTOR;
+        // Rounding division for signed numerator
+        if (ln > 0) {
+            return (ln + _HALF_COMPRESSION_FACTOR) / _COMPRESSION_FACTOR;
+        } else {
+            return (ln - _HALF_COMPRESSION_FACTOR) / _COMPRESSION_FACTOR;
+        }
     }
 
-    function _calculateSpotPriceLn(
+    /**
+     * @dev Restores `value` from logarithmic space. `value` is expected to be the result of a call to `_toLowResLog`,
+     * any other function that returns 4 decimals fixed point logarithms, or the sum of such values.
+     */
+    function _fromLowResLog(int256 value) internal pure returns (uint256) {
+        return uint256(LogExpMath.exp(value * _COMPRESSION_FACTOR));
+    }
+
+    /**
+     * @dev Calculates the logarithm of the spot price of token B in token A.
+     *
+     * The return value is a 4 decimal fixed-point number: use `_fromLowResLog` to recover the original value.
+     */
+    function _calcLogSpotPrice(
         uint256 normalizedWeightA,
         uint256 balanceA,
         uint256 normalizedWeightB,
         uint256 balanceB
-    ) internal pure returns (int256 spotPriceLn) {
-        //Rounding direction does not matter because we are compressing the log result at the end.
-        uint256 spotPrice = balanceA.divDown(normalizedWeightA).divDown(balanceB.divUp(normalizedWeightB));
-
-        spotPriceLn = SignedFixedPoint.ln(spotPrice.toInt256()) / _PRICE_COMPRESSION_FACTOR;
+    ) internal pure returns (int256) {
+        // Max balances are 2^112 and min weights are 0.01, so balance / weight can always be computed.
+        // Rounding direction is irrelevant as we're about to introduce much larger error when converting to log space:
+        // we use divDown because it uses less gas.
```

Comment:
```suggestion
        // The rounding direction is irrelevant as we're about to introduce a much larger error when converting to log
        // space: we use `divDown` because it uses less gas.
```